File: src/Systems/Go/factory_method.go
package main

import "fmt"

type database interface {
	connect()
	query()
}

type postgresDatabase struct{}
func (postgresDatabase) connect() { fmt.Println("PostgreSQL connect") }
func (postgresDatabase) query() { fmt.Println("PostgreSQL query") }

type mySqlDatabase struct{}
func (mySqlDatabase) connect() { fmt.Println("MySQL connect") }
func (mySqlDatabase) query() { fmt.Println("MySQL query") }

type factoryMethod func() database

func useDatabase(create factoryMethod) {
	database := create()
	database.connect()
	database.query()
}

func main() {
	useDatabase(func() database { return postgresDatabase{} })
	useDatabase(func() database { return mySqlDatabase{} })
}

File: src/Systems/Rust/factory_method.rs
trait Database {
    fn connect(&self);
    fn query(&self);
}

struct PostgresDatabase;
impl Database for PostgresDatabase {
    fn connect(&self) { println!("PostgreSQL connect"); }
    fn query(&self) { println!("PostgreSQL query"); }
}

struct MySqlDatabase;
impl Database for MySqlDatabase {
    fn connect(&self) { println!("MySQL connect"); }
    fn query(&self) { println!("MySQL query"); }
}

fn use_database(create: fn() -> Box<dyn Database>) {
    let database = create();
    database.connect();
    database.query();
}

fn create_postgres() -> Box<dyn Database> { Box::new(PostgresDatabase) }
fn create_mysql() -> Box<dyn Database> { Box::new(MySqlDatabase) }

fn main() {
    use_database(create_postgres);
    use_database(create_mysql);
}
